File: app/src/main/java/net/opendasharchive/openarchive/services/webdav/WebDAVLoginActivity.java
package net.opendasharchive.openarchive.services.webdav;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.text.TextUtils;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.view.ContextThemeWrapper;
import androidx.appcompat.widget.Toolbar;

import com.google.android.material.snackbar.Snackbar;
import com.thegrizzlylabs.sardineandroid.Sardine;
import com.thegrizzlylabs.sardineandroid.impl.OkHttpSardine;
import com.thegrizzlylabs.sardineandroid.impl.SardineException;

import net.opendasharchive.openarchive.R;
import net.opendasharchive.openarchive.db.Media;
import net.opendasharchive.openarchive.db.Project;
import net.opendasharchive.openarchive.db.Space;
import net.opendasharchive.openarchive.util.Prefs;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.List;

/**
 * A login screen that offers login via email/password.
 */
public class WebDAVLoginActivity extends AppCompatActivity {

    private final static String TAG = "Login";

    /**
     * Keep track of the login task to ensure we can cancel it if requested.
     */
    private Thread mAuthThread = null;

    // UI references.
    private EditText mNameView, mEmailView, mServerView, mPasswordView;
    private View mProgressView;
    private View mLoginFormView;
    private Space mSpace;

    private Snackbar mSnackbar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        mSpace = null;

        if (getIntent().hasExtra("space")) {
            mSpace = Space.findById(Space.class, getIntent().getLongExtra("space", -1L));
            findViewById(R.id.action_remove_space).setVisibility(View.VISIBLE);
        }
        else {
            mSpace = new Space();
            mSpace.setType(Space.TYPE_WEBDAV);
        }

        // Set up the login form.
        mNameView = findViewById(R.id.servername);
        mEmailView = findViewById(R.id.email);
        mServerView = findViewById(R.id.server);


        if (!TextUtils.isEmpty(mSpace.getName()))
            mNameView.setText(mSpace.getName());

        if (!TextUtils.isEmpty(mSpace.getHost()))
            mServerView.setText(mSpace.getHost());


        if (!TextUtils.isEmpty(mSpace.getUsername()))
            mEmailView.setText(mSpace.getUsername());

        mPasswordView = findViewById(R.id.password);
        mPasswordView.setOnEditorActionListener((textView, id, keyEvent) -> {
            if (id == EditorInfo.IME_ACTION_DONE || id == EditorInfo.IME_NULL) {
                attemptLogin();
                return true;
            }
            return false;
        });

        mLoginFormView = findViewById(R.id.login_form);
        mProgressView = findViewById(R.id.login_progress);

        mSnackbar = Snackbar.make(findViewById(R.id.loginform),"Logging in...",Snackbar.LENGTH_INDEFINITE);

        Intent intent = getIntent();

        if (intent != null && intent.hasExtra("user"))
        {
            String user = intent.getStringExtra("user");
            String password = intent.getStringExtra("password");
            String server = intent.getStringExtra("server");

            mNameView.setText(server);
            mServerView.setText(server);
            mEmailView.setText(user);
            mPasswordView.setText(password);
        }
    }

    @Override
    protected void onPause() {
        super.onPause();


        if (mSpace != null && (!TextUtils.isEmpty(mSpace.getName()))) {
            if (!mNameView.getText().toString().equals(mSpace.getName()))
            {
                mSpace.setName(mNameView.getText().toString());
                mSpace.save();
            }
        }

    }

    /**
     * Attempts to sign in or register the account specified by the login form.
     * If there are form errors (invalid email, missing fields, etc.), the
     * errors are presented and no actual login attempt is made.
     */
    private void attemptLogin() {
        if (mAuthThread != null && mAuthThread.isAlive()) {
            return;
        }

        mSnackbar.show();

        // Reset errors.
        mEmailView.setError(null);
        mPasswordView.setError(null);

        // Store values at the time of the login attempt.
        mSpace.setName(mNameView.getText().toString());
        mSpace.setUsername(mEmailView.getText().toString());
        mSpace.setPassword(mPasswordView.getText().toString());
        mSpace.setHost(mServerView.getText().toString());

        if (TextUtils.isEmpty(mSpace.getName()))
            mSpace.setName(mSpace.getHost());

        boolean cancel = false;
        View focusView = null;

        // Check for a valid password, if the user entered one.
        if (!TextUtils.isEmpty(mSpace.getPassword()) && !isPasswordValid(mSpace.getPassword())) {
            mPasswordView.setError(getString(R.string.error_invalid_password));
            focusView = mPasswordView;
            cancel = true;
        }

        // Check for a valid email address.
        if (TextUtils.isEmpty(mSpace.getUsername())) {
            mEmailView.setError(getString(R.string.error_field_required));
            focusView = mEmailView;
            cancel = true;
        } else if (!isEmailValid(mSpace.getUsername())) {
            mEmailView.setError(getString(R.string.error_invalid_email));
            focusView = mEmailView;
            cancel = true;
        }

        String hostStr = "";
        if (!mSpace.getHost().toLowerCase().startsWith("http"))
        {
            //auto add nextcloud defaults
            hostStr = "https://" + mSpace.getHost() + "/remote.php/dav/";
        }
        else if (!mSpace.getHost().contains("/dav"))
        {
            hostStr = mSpace.getHost() + "/remote.php/dav/";
        }
        mSpace.setHost(hostStr);

        if (cancel) {
            // There was an error; don't attempt login and focus the first
            // form field with an error.
            focusView.requestFocus();
        } else {
            // Show a progress spinner, and kick off a background task to
            // perform the user login attempt.
            mAuthThread = new Thread(new UserLoginTask());
            mAuthThread.start();

        }
    }

    private boolean isEmailValid(String email) {
        return !TextUtils.isEmpty(email);
    }

    private boolean isPasswordValid(String password) {
        return password.length() > 0;
    }

    private Handler mHandlerLogin = new Handler ()
    {
        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);

            switch(msg.what)
            {
                case 0:
                    mSnackbar.dismiss();

                    //success;
                    finish();

                    break;
                case 1:
                default:
                    mSnackbar.dismiss();

                    mPasswordView.setError(getString(R.string.error_incorrect_password));
                    mPasswordView.requestFocus();

                    break;
            }
        }
    };

    /**
     * Represents an asynchronous login/registration task used to authenticate
     * the user.
     */
    public class UserLoginTask implements Runnable {


        @Override
        public void run () {

            if (TextUtils.isEmpty(mSpace.getHost()))
                return;

            try {
                new URL(mSpace.getHost()).toURI();
            } catch (MalformedURLException | URISyntaxException malformedURLException) {
                //not a valid URL
                return;
            }

            StringBuffer siteUrl = new StringBuffer();
            siteUrl.append(mSpace.getHost());
            if (!mSpace.getHost().endsWith("/"))
                siteUrl.append("/");

            Sardine sardine = new OkHttpSardine();
            sardine.setCredentials(mSpace.getUsername(),mSpace.getPassword());

            try {
                try {
                                        sardine.getQuota(siteUrl.toString());
                    mSpace.save();
                    Prefs.setCurrentSpaceId(mSpace.getId());


                    mHandlerLogin.sendEmptyMessage(0);
                } catch (SardineException se) {
                    if (se.getStatusCode() == 401) {
                        //unauthorized
                        Log.e(TAG, "error on login: " + siteUrl.toString(), se);

                        mHandlerLogin.sendEmptyMessage(1);
                    } else {
                        //try again?
                        siteUrl.append("remote.php/dav/");
                        sardine.getQuota(siteUrl.toString());
                        Prefs.setCurrentSpaceId(mSpace.getId());
                        mSpace.save();


                        mHandlerLogin.sendEmptyMessage(0);
                    }
                } catch (IOException e) {

                    //try again?

                    siteUrl.append("remote.php/dav/");
                    sardine.getQuota(siteUrl.toString());
                    Prefs.setCurrentSpaceId(mSpace.getId());
                    mSpace.save();


                    mHandlerLogin.sendEmptyMessage(0);
                }
            }
            catch (SardineException se)
            {
                if (se.getStatusCode() == 401)
                {
                    //unauthorized
                    Log.e(TAG,"unauthorized login: " + siteUrl.toString(),se);

                    mHandlerLogin.sendEmptyMessage(1);
                }
                else
                {
                    Log.e(TAG,"login error: " + siteUrl.toString(),se);

                    mHandlerLogin.sendEmptyMessage(0);
                }
            } catch (IOException e) {

                //nope that is legit an error
                Log.e(TAG,"error on login: " + siteUrl.toString(),e);

                mHandlerLogin.sendEmptyMessage(1);

            }

        }

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        getMenuInflater().inflate(R.menu.menu_login, menu);


        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_sign_in:
                attemptLogin();
                return true;

            // Respond to the action bar's Up/Home button
            case android.R.id.home:
                //NavUtils.navigateUpFromSameTask(this);
                finish();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public void removeProject (View view) {

        DialogInterface.OnClickListener dialogClickListener = new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                switch (which){
                    case DialogInterface.BUTTON_POSITIVE:
                        //Yes button clicked
                        confirmRemoveSpace();
                        finish();
                        break;

                    case DialogInterface.BUTTON_NEGATIVE:
                        //No button clicked
                        break;
                }
            }
        };

        String message = getString(R.string.confirm_remove_space);


        AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(this, R.style.AlertDialogCustom));
        builder.setTitle(R.string.remove_from_app)
                .setMessage(message).setPositiveButton(R.string.action_remove, dialogClickListener)
                .setNegativeButton(R.string.action_cancel, dialogClickListener).show();
    }

    private void confirmRemoveSpace () {
        mSpace.delete();

        List<Project> listProjects = Project.Companion.getAllBySpace(mSpace.getId());

        for (Project project : listProjects)
        {

            List<Media> listMedia = Media.Companion.getMediaByProject(project.getId());

            for (Media media : listMedia)
            {
                media.delete();
            }

            project.delete();
        }

        finish();
    }
}

File: app/src/main/java/net/opendasharchive/openarchive/services/webdav/WebDAVLoginActivity.kt
package net.opendasharchive.openarchive.services.webdav

import android.annotation.SuppressLint
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.text.TextUtils
import android.util.Log
import android.view.*
import android.view.inputmethod.EditorInfo
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.view.ContextThemeWrapper
import com.google.android.material.snackbar.Snackbar
import com.orm.SugarRecord.findById
import com.thegrizzlylabs.sardineandroid.Sardine
import com.thegrizzlylabs.sardineandroid.impl.OkHttpSardine
import com.thegrizzlylabs.sardineandroid.impl.SardineException
import net.opendasharchive.openarchive.MainActivity
import net.opendasharchive.openarchive.R
import net.opendasharchive.openarchive.databinding.ActivityLoginBinding
import net.opendasharchive.openarchive.db.Collection
import net.opendasharchive.openarchive.db.Media
import net.opendasharchive.openarchive.db.Media.Companion.getMediaByProject
import net.opendasharchive.openarchive.db.Project.Companion.getAllBySpace
import net.opendasharchive.openarchive.db.Space
import net.opendasharchive.openarchive.db.SpaceChecker
import net.opendasharchive.openarchive.features.core.BaseActivity
import net.opendasharchive.openarchive.util.Constants
import net.opendasharchive.openarchive.util.Constants.EMPTY_STRING
import net.opendasharchive.openarchive.util.Globals
import net.opendasharchive.openarchive.util.Prefs
import net.opendasharchive.openarchive.util.Prefs.setCurrentSpaceId
import net.opendasharchive.openarchive.util.Utility
import net.opendasharchive.openarchive.util.extensions.isEmailValid
import net.opendasharchive.openarchive.util.extensions.show
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import java.net.URL
import java.util.*

class WebDAVLoginActivity : BaseActivity() {

    private lateinit var mBinding: ActivityLoginBinding
    private lateinit var mSnackbar: Snackbar
    private var mCollection: List<net.opendasharchive.openarchive.db.Collection>? = null
    private var mSpace: Space? = null
    private var mAuthThread: Thread? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE)
        mBinding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(mBinding.root)

        if (intent.hasExtra(Constants.SPACE_EXTRA)) {
            mSpace = findById(Space::class.java, intent.getLongExtra(Constants.SPACE_EXTRA, -1L))
            mBinding.actionRemoveSpace.show()
        } else {
            mSpace = Space()
            mSpace?.type = Space.TYPE_WEBDAV
        }
        initView()
    }


    private fun initView() {
        setSupportActionBar(mBinding.toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        // Set up the login form.
        mSpace?.let { space ->
            if (space.name.isNotEmpty()) {
                mBinding.servername.setText(space.name)
            }
            if (space.host.isNotEmpty()) {
                mBinding.server.setText(space.host)
            }
            if (space.username.isNotEmpty()) {
                mBinding.email.setText(space.username)
            }
        }

        mBinding.password.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE || actionId == EditorInfo.IME_NULL) {
                attemptLogin()
            }
            false
        }

        mSnackbar = Snackbar.make(
            mBinding.loginForm,
            getString(R.string.login_activity_logging_message),
            Snackbar.LENGTH_INDEFINITE
        )

        with(intent) {
            if (hasExtra(Constants.EXTRA_DATA_USER)) {
                val user = getStringExtra(Constants.EXTRA_DATA_USER)
                val password = getStringExtra(Constants.EXTRA_DATA_PASSWORD)
                val server = getStringExtra(Constants.EXTRA_DATA_SERVER)

                mBinding.servername.setText(server)
                mBinding.server.setText(server)
                mBinding.email.setText(user)
                mBinding.password.setText(password)
            }
        }

    }

    /**
     * Attempts to sign in or register the account specified by the login form.
     * If there are form errors (invalid email, missing fields, etc.), the
     * errors are presented and no actual login attempt is made.
     */
    private fun attemptLogin() {
        var focusView: View? = null
        var hostStr: String

        //Reset errors
        mBinding.email.error = null
        mBinding.password.error = null

        // Store values at the time of the login attempt.
        mSpace?.let { space ->
            var cancel = false
            with(mBinding) {
                space.name = servername.text?.toString() ?: EMPTY_STRING
                space.username = email.text?.toString() ?: EMPTY_STRING
                space.password = password.text?.toString() ?: EMPTY_STRING
                space.host = server.text?.toString() ?: EMPTY_STRING

                ///added validation for username, password and web server field
                if (space.name.isEmpty()) {
                    space.name = space.host
                }
                if (mBinding.server.text!!.isEmpty()) {
                    mBinding.server.error = getString(R.string.error_field_required)
                    focusView = mBinding.server
                    cancel = true
                } else if (space.username.isEmpty()) {
                    email.error = getString(R.string.error_field_required)
                    focusView = mBinding.email
                    cancel = true
                } else if (!space.username.isEmailValid()) {
                    email.error = getString(R.string.error_invalid_email)
                    cancel = true
                } else if (space.password.isEmpty()) {
                    password.error = getString(R.string.error_field_required)
                    focusView = mBinding.password
                    cancel = true
                }
            }

            hostStr = updateUserProvidedServerUrlToCorrectPath(space)
            space.host = hostStr

            if (cancel) {
                // There was an error; don't attempt login and focus the first
                // form field with an error.
                focusView?.requestFocus()
            } else {
                // Show a progress spinner, and kick off a background task to
                // perform the user login attempt.
                mSnackbar.show()
                mAuthThread = Thread(UserLoginTask())
                mAuthThread?.start()
            }
        }
    }

    private fun updateUserProvidedServerUrlToCorrectPath(
        space: Space,
    ): String {
        var fullUrl = EMPTY_STRING
        if (!space.host.lowercase(Locale.ROOT).startsWith(Constants.PREFIX_HTTP)) {
            //auto add nextcloud defaults
            fullUrl = "${Constants.PREFIX_HTTPS}${space.host}${Constants.REMOTE_PHP_ADDRESS}"
        } else if (!space.host.contains(Constants.DAV)) {
            fullUrl = "${space.host}${Constants.REMOTE_PHP_ADDRESS}"
        }
        return fullUrl
    }

    private val mHandlerLogin: Handler = @SuppressLint("HandlerLeak")
    object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            when (msg.what) {
                0 -> {
                    mSnackbar.dismiss()
                    finish()
                }
                1 -> {
                    mSnackbar.dismiss()
                    mBinding.password.error = getString(R.string.error_incorrect_password)
                    mBinding.password.requestFocus()
                }
                2 -> {
                    mSnackbar.dismiss()
                    Toast.makeText(
                        this@WebDAVLoginActivity,
                        getString(R.string.login_please_use_username),
                        Toast.LENGTH_LONG
                    ).show()
                }
                else -> {
                    mSnackbar.dismiss()
                    mBinding.password.error = getString(R.string.error_incorrect_password)
                    mBinding.password.requestFocus()
                }
            }
        }
    }

    /**
     * Represents an asynchronous login/registration task used to authenticate
     * the user.
     */
    inner class UserLoginTask : Runnable {
        override fun run() {

            mSpace?.let { space ->
                if (TextUtils.isEmpty(space.host)) return
                try {
                    URL(space.host).toURI()
                } catch (malformedURLException: Exception) {
                    //not a valid URL or incorrect syntax error
                    return
                }
                val sardine: Sardine = OkHttpSardine()
                sardine.setCredentials(space.username, space.password)
                try {
                    try {
                        if (loginUserIntoWebDav(space.host, space.username, space.password)) {
                            if (Space.getSpaceForCurrentUsername(
                                    space.username,
                                    Space.TYPE_WEBDAV,
                                    space.host
                                ) == 0
                            ) {
                                space.save()
                                setCurrentSpaceId(space.id)
                                finishAffinity()
                                startActivity(
                                    Intent(
                                        this@WebDAVLoginActivity,
                                        MainActivity::class.java
                                    )
                                )

                            } else {
                                showLoginIntoWebDavErrorMessage()
                            }
                        } else {
                            mHandlerLogin.sendEmptyMessage(1)
                        }
                    } catch (se: SardineException) {
                        when (se.statusCode) {
                            401 -> mHandlerLogin.sendEmptyMessage(1)
                            404 -> mHandlerLogin.sendEmptyMessage(2)
                            else -> {}
                        }
                    } catch (exception: IOException) {
                        exception.printStackTrace()
                    }
                } catch (sardineException: SardineException) {
                    when (sardineException.statusCode) {
                        401 -> mHandlerLogin.sendEmptyMessage(1)
                        404 -> mHandlerLogin.sendEmptyMessage(2)
                        else -> mHandlerLogin.sendEmptyMessage(0)
                    }
                } catch (exception: IOException) {
                    mHandlerLogin.sendEmptyMessage(1)
                }
            }
        }
    }

    private fun showLoginIntoWebDavErrorMessage() {
        runOnUiThread {
            mSnackbar.dismiss()
            Toast.makeText(
                this@WebDAVLoginActivity,
                getString(R.string.login_you_have_already_space),
                Toast.LENGTH_LONG
            ).show()
        }
    }

    fun loginUserIntoWebDav(fullUrl: String, username: String, password: String): Boolean {
        val okHttpBaseClient = Utility.generateOkHttpClient(this)
        val request: Request = Request.Builder()
            .url(fullUrl)
            .method("GET", null)
            .addHeader("OCS-APIRequest", "true")
            .addHeader("Accept", "application/json")
            .build()
        val response: Response = okHttpBaseClient.newCall(request).execute()
        Prefs.putString(Globals.PREF_NEXTCLOUD_USER_DATA, response.body?.string())
        return response.code == 200
    }

    fun removeProject(v: View) {
        val dialogClickListener = DialogInterface.OnClickListener { _, which ->
            when (which) {
                DialogInterface.BUTTON_POSITIVE -> {
                    confirmRemoveSpace()
                    finish()
                }
                DialogInterface.BUTTON_NEGATIVE -> {}
            }
        }
        val message = getString(R.string.confirm_remove_space)
        val builder = AlertDialog.Builder(ContextThemeWrapper(this, R.style.AlertDialogCustom))
        builder.setTitle(R.string.remove_from_app)
            .setMessage(message).setPositiveButton(R.string.action_remove, dialogClickListener)
            .setNegativeButton(R.string.action_cancel, dialogClickListener).show()
    }

    private fun confirmRemoveSpace() {
        mSpace?.let { space ->
            space.delete()
            space.id?.let {
                val listProjects = getAllBySpace(it)
                listProjects?.forEach { project ->
                    mCollection = Collection.getCollectionById(project.id)
                    mCollection?.forEach { collection ->
                        collection.delete()
                    }
                    mCollection = null

                    val listMedia = getMediaByProject(project.id)
                    listMedia?.forEach { media ->
                        media.delete()
                    }
                    project.delete()
                }
            }
            SpaceChecker.navigateToHome(this)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_login, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_sign_in -> {
                attemptLogin()
                return true
            }
            android.R.id.home -> {
                finish()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }
}